### ISSUE TITLE
How to use Native MTP?

### ISSUE THREAD

**Issue by NirvanaCh7** [jundot/omlx#1631]
I tried running mlx-community/Qwen3.6-35B-A3B-OptiQ-4bit.

I noticed there is a mtp.safetensors file in the model folder, but the model configuration shows:
```
Config declares MTP layers, but the converted weights are missing mtp.* tensors.
Please re-convert from HuggingFace using a converter that preserves MTP weights.
```
As a result, the Native MTP option is disabled. Maybe it's a bug?

I also found some draft models:

mlx-community/Qwen3.6-35B-A3B-MTP-bf16

Can we use this draft model? (as a new feature)

Additionally, I tested TurboQuant KV Cache, but I did not observe any significant reduction in peak memory usage during benchmarking.
Am I misunderstanding something?

**Comment by zviratko** [comment]
most models had MTP stripped
some claim MTP support but either don't have it or it is incompatible with oMLX

draft model (if you're talking about SpecPrefill) would be something like Qwen3.6-0.8B (nothing special, just small with the same tokens), not just a bf19 version of one
or DFlash (in z-lab/, named DFlash :))

IMO turboquant doesn't work with all models (does it?) and it quantized the cache only after all blocks have been prefilled, so it doesn't help with peak memory usage for long contexts, it never did anything in my benchmarking, maybe someone will shed more light on when it is supposed to work

**Comment by NirvanaCh7** [comment]
> most models had MTP stripped some claim MTP support but either don't have it or it is incompatible with oMLX
> 
> draft model (if you're talking about SpecPrefill) would be something like Qwen3.6-0.8B (nothing special, just small with the same tokens), not just a bf19 version of one or DFlash (in z-lab/, named DFlash :))
> 
> IMO turboquant doesn't work with all models (does it?) and it quantized the cache only after all blocks have been prefilled, so it doesn't help with peak memory usage for long contexts, it never did anything in my benchmarking, maybe someone will shed more light on when it is supposed to work

So, the MTP of OptiQ is unique, and it requires an optiq backend to operate?
Or it's better to use the oQ quantified model with MTP ?

mlx-community/Qwen3.6-35B-A3B-MTP-bf16 is a 0.8B MTP drafter checkpoints -- says mlx-community. I don't know if it should be filling into "SpecPrefill" option.

**Comment by NirvanaCh7** [comment]
https://github.com/jundot/omlx/discussions/454#discussioncomment-16364127

**Comment by dhanna11** [comment]
I ran into this issue as well. The sidecar MTP file is non-standard and is designed to be used in the OptiQ backend. Easier just to quantize it using the oQ quantizer or pull down a model with the actual MTP layers still present.

**Comment by maxanstey-meridian** [comment]
If anyone else is having this issue and wants a "workaround" use Unsloth's UD version. It's not MTP, it's not OptiQ, but it's fast and it's close https://huggingface.co/unsloth/Qwen3.6-35B-A3B-UD-MLX-4bit

If you want to use MTP, etc. without oMLX, it's: `llama-server -hf unsloth/Qwen3.6-35B-A3B-MTP-GGUF:UD-Q4_K_XL -ngl 999 -fa on --spec-type draft-mtp --jinja -c 102400`

**Comment by tomtastic** [comment]
I'm also wanting to use Native MTP, but the mlx-community models don't seem to work.
I guess I want a native single MLX model with MTP weights included, does that exist in 8-bit quantisation?

https://huggingface.co/mlx-community/Qwen3.6-27B-bf16
and
https://huggingface.co/mlx-community/Qwen3.6-27B-MTP-bf16

<img width="1330" height="190" alt="Image" src="https://github.com/user-attachments/assets/5f61bd31-b68c-4cba-aa0e-f8cde2235028" />

**Comment by brjen** [comment]
Adding a data point for anyone going the mlx-optiq route to get native MTP (since the mtp.safetensors sidecar is OptiQ-specific, as @dhanna11 said): on mlx-optiq 0.4.25, `--mtp` combined with `--kv-bits` is a **silent no-op** — KV quant keeps the server on the batch generator, MTP only hooks the sequential path, so the engine never attaches. No warning, byte-identical output, 0% speedup. The MTP engine also prefills the whole prompt in one forward pass (materializing logits for every position), which Metal-OOMs around a 3K-token prompt on a 27B.

I wrote up the three bugs + a drop-in wrapper (forces sequential, forwards kv_bits, chunks prefill) and a script that builds the OptiQ-style MTP model dir from the mlx-community MTP head repos: https://github.com/brjen/optiq-mtp-fix

Measured on Qwen3.8-27B 4-bit, M5 Air 24 GB: greedy decode 7.7–8.2 → 11.6–12.1 tok/s (depth 1), 60K prompt prefill clean, 64K context with kv8.

**Comment by zviratko** [comment]
I found that Optiq-4bit is usually comparable to oQ5e in size and oQ5e is very slightly "smarter" as far as benchmarks can tell (not always the case for oQ4e) and the oQ quants are faster and have working MTP, so you can try - *-oQ5e-mtp or *-oQ4e-mtp quants instead.

(I haven't compared Qwen3.8 quants but this was my experience with Qwen3.6)

**Comment by brjen** [comment]
Thanks @zviratko — worth a shot. I'll benchmark Qwen3.8-27B oQ5e-mtp (and oQ4e-mtp) against the OptiQ+wrapper path on an M5 Air 24 GB — decode tok/s, long-context prefill at 60K, and peak wired memory — and report back here.

**Comment by brjen** [comment]
Followed up on @zviratko's suggestion — benchmarked **Qwen3.8-27B oQ4e-mtp** (Jundot's quant) and **oQ5e-mtp** under oMLX 0.6.3rc1 on an M5 Air 24 GB, vs my OptiQ-4bit + wrapper setup:

| | oMLX oQ4e-mtp | mlx-optiq 4bit + MTP fix |
|---|---|---|
| model size | 16.6 GB | ~15 GB |
| load → first token | **6.7 s** (impressive) | 30–60 s |
| greedy decode | 8.2 tok/s | 7.7–8.2 plain, **11.6–12.1 MTP d1** |
| 4K prompt prefill | ❌ aborted — peak 20.1 GB vs 19.9 GB watermark | ✅ |
| 32K / 60K prefill | ❌ 400 / not reachable | ✅ 60K clean (chunked prefill, kv8) |

Notes for the 24 GB crowd:
- Decode exactly matched my *non*-MTP baseline, and the serve log shows the model was discovered as `type: vlm` and loaded on `VLMBatchedEngine` — MTP patches applied at startup, but no draft/accept activity in the logs. Looks like the same failure shape as the mlx-optiq bug above: MTP hooks the sequential path and the batched engine silently skips it. If there is a way to force the sequential engine for a text-only -mtp model via `omlx serve`, I could not find a flag for it — happy to re-test if someone knows the knob.
- Prefill materializes enough transient memory that even a 4K prompt blows past the hard watermark with a 27B resident — at the default 18 GB guard *and* raised to the 21 GB iogpu cap. Chunked prefill is what saves the optiq path at 60K.
- **oQ5e-mtp (20.3 GB) is a hard no on 24 GB**: it loads (18.3 GB), then the first request hit 21.2 GB and hard-froze the machine (panic + reboot). Also the 3.6-era size note does not carry to 3.8: OptiQ-4bit is ~15 GB here, well under oQ5e's 20.3 GB.

So on a 24 GB machine the OptiQ + chunked-prefill route stays the best I have found for this model; the oQ-mtp quants probably need 32 GB+ to stretch their legs. The 6.7 s model load and the SSD KV-cache tier are genuinely nice though — I will revisit oMLX on a bigger box.
